checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1602 @@
+    if (mustCallValue == null || mustCallValue.isEmpty()) {
+      return;
+    }
+
+    boolean mustCallSatisfied = false;
+    for (LocalVarWithTree localVarWithTree : fact) {
+
+      // sometimes the store is null!  this looks like a bug in checker dataflow.
+      // TODO track down and report the root-cause bug
+      CFValue lhsCFValue = cmStore != null ? cmStore.getValue(localVarWithTree.localVar) : null;
+      AnnotationMirror cmAnno = null;
+
+      if (lhsCFValue != null) { // When store contains the lhs
+        for (AnnotationMirror anno : lhsCFValue.getAnnotations()) {
+          if (AnnotationUtils.areSameByName(
+              anno, "org.checkerframework.checker.calledmethods.qual.CalledMethods")) {
+            cmAnno = anno;
+          }
+        }
+      }
+      if (cmAnno == null) {
+        cmAnno =
+            typeFactory
+                .getAnnotatedType(localVarWithTree.localVar.getElement())
+                .getAnnotationInHierarchy(typeFactory.top);
+      }
+
+      if (calledMethodsSatisfyMustCall(mustCallValue, cmAnno)) {
+        mustCallSatisfied = true;
+        break;
+      }
+    }
+
+    if (!mustCallSatisfied) {
+      LocalVarWithTree firstlocalVarWithTree = fact.iterator().next();
+      if (!reportedMustCallErrors.contains(firstlocalVarWithTree)) {
+        if (!checker.shouldSkipUses(TreeUtils.elementFromTree(firstlocalVarWithTree.tree))) {
+          reportedMustCallErrors.add(firstlocalVarWithTree);
+          checker.reportError(
+              firstlocalVarWithTree.tree,
+              "required.method.not.called",
+              formatMissingMustCallMethods(mustCallValue),
+              firstlocalVarWithTree.localVar.getType().toString(),
+              outOfScopeReason);
+        }
+      }
+    }
+  }
+
+  /**
+   * Increment the -AcountMustCall counter.
+   *
+   * @param node the node being counted, to extract the type
+   */
+  private void incrementNumMustCall(Node node) {
+    if (checker.hasOption(ResourceLeakChecker.COUNT_MUST_CALL)) {
+      TypeMirror type = node.getType();
+      incrementMustCallImpl(type);
+    }
+  }
+
+  /**
+   * Increment the -AcountMustCall counter.
+   *
+   * @param elt the elt being counted, to extract the type
+   */
+  private void incrementNumMustCall(Element elt) {
+    if (checker.hasOption(ResourceLeakChecker.COUNT_MUST_CALL)) {
+      TypeMirror type = elt.asType();
+      incrementMustCallImpl(type);
+    }
+  }
+
+  /**
+   * Shared implementation for the two version of countMustCall. Don't call this directly.
+   *
+   * @param type the type of the object that has a must call obligation
+   */
+  private void incrementMustCallImpl(TypeMirror type) {
+    // only count uses of JDK classes, since that's what we report on in the paper
+    String qualifiedName = TypesUtils.getTypeElement(type).getQualifiedName().toString();
+    if (!qualifiedName.startsWith("java")) {
```

Comment:
This same test occurs in both `reportError` and `incrementMustCallImpl`.  It is important enough that it should be abstracted out and documented, to avoid them getting out of sync.